docs: add OEP-0058 - Translations Management (#367)
openedx/open-edx-proposals: oeps/architectural-decisions/oep-0058-arch-translations-management.rst

  rights on Transifex in order to function. Admin rights give access to private/sensitive
  information as well as the ability to permanently delete translation and configuration
  files. At some point, the login to the edx-transifex-bot user was lost, and without
  access to the scripts that the bot uses to function, this edx-transifex-bot is a
  security issue we cannot control or debug.
* Edx-transifex-bot is a black box: Login/scripts for edx-transifex-bot cannot be found
  on GitHub or on wiki/secrets pages, and it is difficult to observe what work it is
  supposed to do, or whether it is doing it correctly.
* The translations for the Open edX Maple Release were never uploaded to Transifex,
  because the automation handled by the edx-transifex-bot never uploaded it.
* The translations for the Open edX Nutmeg Release were nearly not uploaded because the
  edx-transifex-bot did not automatically upload it. In addition, the script for
  uploading the translations did not work, and the script had to be tinkered for the new
  Transifex API in order to upload it in time for release. The files uploaded were still

>   not in the preferred format because the new version of the Transifex API does not have

Author: In what way were they not in the preferred format, and how did you know that?

>   not in the preferred format because the new version of the Transifex API does not have

Author: This bullet point has been split into two bullet points that address the technology and the complex process separately. And a link to the Nutmeg Release issue has been added for more context.

  all the same features as the previous API version.

>   * Running the script to upload translations requires Transifex admin rights, and there

Author: I've absorbed the sub-bullet to the parent bullet.

    are few people with Admin rights to Transifex and knowledge of the Transifex API. We
    should use the solution provided by Transifex to solve this type of problem.

Rationale
*********

* This is an upgrade of a system we use regularly, but do not want to have to maintain
  regularly.
* Upgrading from a bot (machine user) to an app/workflow is recommended by GitHub and
  makes the translation process more open source. 
* The GitHub Transifex Integration App is developed and supported by Transifex
* The Github Transifex App is very simple to configure and with many options. We can set
  Transifex Projects to automatically upload/download translation files from a repository
  once the translations are reviewed and accepted.

> * Transifex only allows a one-to-one relationship between repositories and Transifex

Author: I've split the bullets into the two sections.

  Projects. Organizing all of the translation files into one repository and one Transifex

>   Project has a lower cost: we pay by the project so we end up paying less, and by

Reviewer: We do pay for Transifex, though we have a special agreement as a non-profit/open-source project.  That said, I'm not sure about whether this proposal would cost less or whether cost should be a first class consideration in our architecture for translations.

>   Project has a lower cost: we pay by the project so we end up paying less, and by

Reviewer 2: Agree re: slack that monetary cost should be dropped as a consideration

>   Project has a lower cost: we pay by the project so we end up paying less, and by

Author: Point out the human cost of managing the translation files from many repos on Transifex and remove the monetary capital. "We estimate that it takes X hours per month to manage this, and this will reduce it by Y%."

  putting all translation files in the same place, it is easier to track translation
  progress, and debug translation issues.
* A repository that only contains text/binary files, and uses branches to separate
  translations related to Open edX releases can make all interactions with translations
  very quick and simple due to the ability to clone the branch of a specific release,
  with translations organized by repository name.
* By using an app that is maintained by Transifex the organization, we reduce the
  maintenance burden and are more future proof of changes they might make since they
  maintain both the API and the github App.

Proposed Implementation
***********************

> Move translation files to a new repo

Reviewer 2: Your other headers capitalize all words so making this one consistent.

```suggestion
Move Translation Files to a New Repo
```

====================================

Translation files (of types .mo and .po) currently exist amongst the code/documentation

> they translate. We will move these Translation files from being amongst the

Reviewer 2: I'm not sure I understand why "Translation files" is a proper noun... it reads oddly. My suggestion is to change all of these to "translation files".

Author: Translation Files has been demoted to translation files where appropriate.

code/documentation to their own repo. For easier reintegration, Translation files will be

> kept in the same directory structure as the code/documentation they translate.

Reviewer 2: can you provide an example? eg a sentence like,

For example, files currently found in `repo_name/path/to/file` would now exist in the new repo at path `new_path/to/file`

Author: Added an example.

> Add GitHub Transifex App to openedx organization

Reviewer 2: ```suggestion
Add GitHub Transifex App to openedx Organization
```

================================================

Transifex's GitHub app will need to be added to the openedx GitHub organization in order
to grant the app permissions to push/pull the Translation files. Currently, we manage the
push/pull permissions for the edx-transifex-bot through a number of groups. The Transifex
GitHub app is granted permission on a repository basis, and by moving all the

> Translations to a single repository we reduce the number of user groups we will have to

Reviewer 2: Similarly I don't understand why "Translations" is capitalized.

Do we have to install the GH app on every repo, not just over the org? If it's per-repo that's going to be painful.

We won't need any special user groups, right? So "reduce the number of user groups" could be "eliminate separate translations user groups"?

manage.

> Connect i18n repository to Transifex via app

Reviewer 2: You haven't defined `i18n repository` in this document yet (and I don't think you ever do). For clarity, I'd say:

```suggestion
Connect the New Translation Repository to Transifex
```

============================================

The Transifex webapp accepts configuration files for each Transifex project. By
connecting the single repository containing all Translation files, we only need to make a
single configuration file that allows the Transifex GitHub app to manage the Translation

> files. Based on the Translation Working Group's instruction, we can set parameters that

Reviewer 2: what instruction would they be providing?

Author: They would provide instruction on acceptable translation/review levels before translation files are pushed to the new repo.

automatically push and pull Translation files.

Copy Transifex Memory
=====================

As a last step we can save all the progress the Open edX translators have accomplished by
copying the Transifex Memory, the auto-translation feature that allows for Projects with
similar strings to be automatically translated, from the old projects to this new one. By
moving all the Translation Files to the same repository we can increase the reach of the
Transifex Memory feature to help translate similar strings across the entire

> code/documentation base.

Reviewer 2: wait, are we making a new transifex project, and if so, will existing translators/reviewers maintain their access?

Author: We are making a new transifex project to track the new repo. Existing translations/reviewers will be granted access to this new project given their involvement in a Transifex project that is being made redundant by this OEP.

This has been added to this paragraph.

Impacts
*******

Impact on Translators
=====================

As we approach the end of the translation upgrade process, we will need to tactically
move from multiple transifex projects to a single project.  This will require
coordination with our translators to ensure that moving forward they are providing
translations in the right place.

Impact on Site Operators
========================

Currently the translation files for any given service or library is stored at the same

> place as the code, this has generally simplified the deployment story in the past.  With

Reviewer 2: ```suggestion
place as the code, which has generally simplified the deployment story in the past.  With
```

this change, the translations files will move to their own repository.  As we deprecate
the old translations files, the relevant deployment tooling will need to be updated to
pull down the translations from the new repository as a part of the deployment process.
This will impact both the old Ansible based tooling as well as any new docker based
tooling.

Impact on Developers
====================

While it won’t directly impact the day-to-day workflow of developers, due to the same

> reasons that we impact site operators(new translations location), we will have to update

Reviewer 2: ```suggestion
reasons that we impact site operators(new translations location), we will have to update
```

> development tools as well.

Reviewer 2: Will the process of creating a new repo be the same w/r/t setting up translations tooling/getting the integration working?

Author: We will need new instructions for how to enable translations for a new service/repo when it comes online.

Author: This has been updated.

Locations
*********

Dumps of the translation/localization files from Transifex for the Open edX Release

> project already exist in a repository with the name of openedx/openedx-i18n. A new

Reviewer 2: is openedx-i18n going to still exist alongside openedx-translations?

Author: I believe Regis still uses this repo for Tutor. I imagine after the change there will be some re-evaluations on whether its still needed or not.

Author: The new openedx-translations repo will obviate the need for the openedx-i18n repo. Updating tutor to use the new translations repo and archiving the i18n repo will be part of the migration process.

> repository named openedx/openedx-translations, will be similarly structured, but contain

Reviewer 2: ```suggestion
repository named openedx/openedx-translations will be similarly structured, but it will contain
```

the translation files for all repositories within openedx. The GitHub Transifex app will
be installed in the openedx organization. Similar to how the Build-Test-Release Working
Group creates a new branch for each new named release of edx-platform, translation
releases can also be kept in branches corresponding to edx-platform releases.

Rejected Alternatives
*********************

Rewriting the Current Tooling for the New API
=============================================

The source code for the edx-transifex-bot is missing. We could rewrite the current
tooling to try to solve the problems encountered in the last two Open edX releases and
upgrade to the new API, but this approach would require a full rewrite, potentially more
expensive than doing the rewrite in a way that Transifex more cleanly supports. It should
also be mentioned that GitHub discourages the use of bots and separate bot accounts; they
strongly recommend using GitHub Apps.

> Making a Transifex Project for each Repository

Reviewer 2: ```suggestion
Making a Transifex Project for Each Repository
```

==============================================

> As translation support is provided for more repos, the total cost will also increase. A

Reviewer 2: I read "cost" as monetary, so the next two sentences were jarring. I might suggest:

```suggestion
As translation support is provided for more repos, the effort to maintain the translations infrastructure increases. A
```

Transifex Project houses the content to be translated and needs to be created before any
content can be added for translation. Transifex Projects can only support 1 GitHub
repository each and need to be maintained separately. Maintaining a Transifex Project
involves adjusting configurations, adding new languages, assigning translators to
projects, or any other miscellaneous irregular tasks that would be time-consuming at a

> larger scale. If we adda a Transifex Project, each Transifex Project will need to be

Reviewer 2: ```suggestion
larger scale. If we add a Transifex Project, each Transifex Project will need to be
```

Impact on Developers
====================

While it won’t directly impact the day-to-day workflow of developers, due to the same
reasons that we impact site operators (new translations location), we will have to update
development tools as well. In addion, we will create new instructions for developers on
how to enable translations for a new service/repo when it comes online.

Final State
***********

Repositories that generate translation files will have their translation files generated
and committed via a pull request to the openedx-translation repository via a github
workflow. Once the translation files from edx-platform and other repositories are moved
to the openedx-translations repository, the `Transifex GitHub App`_ will link a Transifex
project of a name such as "Open edX Translations" to the openedx-translations repository.
A `Transifex GitHub Integration configuration file`_ naming the files that are to be
translated and the trigger that pulls translation files back into will be created in the
openedx/translations repository. This link will allow for the `Transifex GitHub App`_ to
automatically manage the push/pull of the translation files without the need for human

> intervention.

Author: Distributed this section as detailed above

When it comes time to cut an Open Release, a new branch will be formed in the
openedx-translations repository for this release. This new branch will correspond with
other release branches such as those found in edx-platform. It is my recommendation that
a tool in the form of a python library is written to enable the placement of the
translation files kept in openedx-translations into the repositories the translation
files are formed from. This tool will manage the placement of translation files through

> an editable translation tool configuration file kept in the repositories that have

Reviewer 2: You say "editable translation tool configuration file" a few times in a row. Could we give it a short, succinct name that is easier to read and more memorable? eg the first time you could say, "an editable translation tool configuration file (`translation-config.yml`) and refer to it as such in the next few sentences?

Also, why does "tool" need to be in the name?

  translations related to Open edX releases can make all interactions with translations
  very quick and simple due to the ability to clone and sparse-checkout the branch of a
  specific release and the directory (repository name) with translation files.

Proposed Implementation
***********************

Move Translation Files to a New Repo
====================================

Translation files (of types .mo and .po) currently exist amongst the code/documentation
they translate. We will move these translation files from being amongst the
code/documentation to their own repo. For example, a translation file for the openedx
repository `edx-platform`_ located at
``edx-platform/conf/locale/en/LC_MESSAGES/django.po`` would be moved to the new
repository with the name openedx-translations_ located at
``openedx-translation/edx-platform/conf/locale/en/LC_MESSAGES/django.po``. For easier
reintegration, translation files will be kept in the same directory structure as the
code/documentation they translate.

Repositories that generate translation files will have their translation files generated
and committed via a pull request to the openedx-translation repository via a GitHub

> workflow. Once the translation files from edx-platform and other repositories are moved

Author: Its almost exactly like that. Checkout a run of the github action: https://github.com/openedx/openedx-translations/actions/runs/3415232508

That's a good point about translation changes made to release branches.
Though the above is just the current iteration of the action, in the future we can make the action run on the latest release branch as well to keep everything updated while it is supported.

> workflow. Once the translation files from edx-platform and other repositories are moved

Reviewer 3: How will the release branch logistics work exactly? I was worried because the [Transifex docs](https://help.transifex.com/en/articles/6265125-github-via-transifex-ui#h_de29ef5cec) state that "the GitHub integration works on top of a single branch."

> workflow. Once the translation files from edx-platform and other repositories are moved

Author: This is a good topic for [Follow-up Work](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3578494977/Translations+Management+Design+Implementation). 

As a first stab: when a new release is cut, a new branch of the translations corresponding to the release branch will also be made. And then a new project can track that branch on Transifex. We can update the GitHub Action to run for multiple branches to keep the translation source files for the new release up to date. And with every new release, the old release's transifex project will be archived, but the translations will still be available on GitHub in the old release's branch. 

The first outstanding issue to this approach would be making sure that Translators have access to this new project as well. It will be up to the Translations WG to manage the latest translation files for the entire platform as well as minor fixes to the latest release.

> workflow. Once the translation files from edx-platform and other repositories are moved

Reviewer 3: I am wildly guessing here as I'm not involved in BTR, but I had assumed current state is that translations are frozen at the time the release is cut, and any needed changes are made directly to the translation files in the release's branch for that repo.

(Not saying this is what we should do in the future, just wanted to make sure I am on the right page.)

It might also be nice to include a way for `openedx-atlas` to sideload translation files for the correct `openedx-translation` named release branch being deployed/developed on. Not sure how that would work. A `--branch` parameter? Some sort of auto-detection?

> workflow. Once the translation files from edx-platform and other repositories are moved

Reviewer 4: > It might also be nice to include a way for `openedx-atlas` to sideload translation files for the correct `openedx-translation` named release branch being deployed/developed on. Not sure how that would work. A `--branch` parameter? Some sort of auto-detection?

the readme on https://github.com/openedx/openedx-atlas documents 2 ways to specify which branch to pull:
* using `--config` and specifying a branch in a config file
* directly specifying the branch using `-b`/`--branch`

=====================

As we approach the end of the translation upgrade process, we will need to tactically
move from multiple Transifex projects to a single project. This will require coordination
with our translators to ensure that moving forward they are providing translations in the
right place.

Impact on Site Operators
========================

Currently the translation files for any given service or library is stored at the same
place as the code, which has generally simplified the deployment story in the past. With
this change, the translations files will move to their own repository. As we deprecate
the old translations files, the relevant deployment tooling will need to be updated to
pull down the translations from the new repository as a part of the deployment process.
This will impact both the old Ansible based tooling as well as any new docker based
tooling.

Impact on Developers
====================

While it won’t directly impact the day-to-day workflow of developers, due to the same
reasons that we impact site operators (new translations location), we will have to update
development tools as well. In addition, we will create new instructions for developers on

> how to enable translations for a new service/repo when it comes online.

Reviewer 5: > While it won’t directly impact the day-to-day workflow of developers

Isn't that the translation will no longer exists on their repo, thus to test or get translation at development time, developer would need to run `openedx-atlas` ?  So I assume that would **impact**  

Secondly, the degree of the impact which is not known yet... it seems every repo suppose to have the `atlas.yaml` file and install the new cli tool, I would argue if it would be possible for that to release itself in binary format, so 1)developer wouldn't have to think about their python version when they want to install it, which can be useful spefically in MFEs development..., i.e. _an MFE doesn't need to consider python env if the library has a binary release_ 2)Which can can make CI/CD workflow more effiecent as well.  

Note: _second comment might be releavnt in this scope I just wanted to raise because I think it can be helpful_

Reviewer 6: How will developers receive errors from the source translation extract/compile step that may happen between openedx repos and openedx-translations? Currently, for example, if I misformat a translation's pluralization in edx-platform python, I can see [an error from django's compilemessages command](https://docs.djangoproject.com/en/4.1/topics/i18n/translation/#pluralization-var-notes), and CI ensures I have to repair it before the PR will merge.

Can we keep these compilation linting checks in place in CI even though the generated files are discarded from the repo? And can we encourage this practice on new repos too, including MFEs?

Author: @pomegranited 
I'm not familiar with this CI that checks the formatting of translations. Could you link it here? I agree that it should not be changed since its very useful. I also agree that this practice would benefit MFEs as well, though I'm less familiar with how that would work.

Reviewer 6: @Carlos-Muniz  Ah.. looks like there aren't any CI checks for missing changes to source translations. I tested this by submitting https://github.com/openedx/edx-platform/pull/31276, to see whether CI would fail, and it didn't.

I could have sworn there were (using [`make detect_changed_source_translations`](https://github.com/openedx/edx-platform/blob/605a0a66e72343ef61c1a5310ed9acbbbf34915a/Makefile#L60-L61) or perhaps [`paver i18n*`](https://github.com/openedx/edx-platform/blob/master/pavelib/i18n.py)?), but I'm likely mis-remembering. Maybe the updated translation extractions are handled by the translations bot instead of CI?

Reviewer 7: In some repositories we have that check as a part of the `quality` step

Locations
*********

Dumps of the translation/localization files from Transifex for the Open edX Release

> project already exist in a repository with the name of openedx/openedx-i18n. A new

Reviewer 8: If the new openedx-translations repo contains strings for all languages and repos and it follows the regular minor/major release schedule, then yes we will be able to get rid of the openedx-i18n repo. I'm looking forward to it :)

repository named openedx/openedx-translations will be similarly structured, but it will
contain the translation files for all repositories within openedx. The
`Transifex GitHub App`_ will be installed in the openedx organization. Similar to how the
Build-Test-Release Working Group creates a new branch for each new named release of
edx-platform, translation releases can also be kept in branches corresponding to

> edx-platform releases.

Author: I originally left it as "can" instead of "will" because I wanted to leave it open to alternatives. If it works with the automatic script that cuts release branches, I am more confident that this is the right way. :+1:

Rejected Alternatives
*********************

Rewriting the Current Tooling for the New API
=============================================

> The source code for the edx-transifex-bot can be found in `ecommerce-scripts`_. We could

Author: This OEP-58 will limit the actions the edx-transifex-bot, but we cannot fully deprecate the bot because it would be instead used as a virtual user that will act as the point of contact in the github action that generates translation source files and the GitHub Transifex App.

> The source code for the edx-transifex-bot can be found in `ecommerce-scripts`_. We could

Reviewer 4: just to add context here, if you look through https://github.com/openedx/openedx-translations/commits/main you'll see commits by `edx-transifex-bot` starting with `chore: add extracted translation source files`

OEP-58: Translations Management
###############################

.. list-table::
   :widths: 25 75

   * - OEP
     - :doc:`OEP-58 <oep-0058-arch-translations-management>`
   * - Title
     - Translations Management
   * - Last Modified
     - 2022-11-07
   * - Authors
     - 
       * Carlos Muniz <cmuniz@tcril.org>
       * Feanil Patel <feanil@tcril.org>
       * Sarina Canelake <sarina@tcril.org>
   * - Arbiter
     - Ned Batchelder <ned@edx.org>
   * - Status

>      - Under Review

Reviewer 9: Based on the [Provisional definition](https://open-edx-proposals.readthedocs.io/en/latest/processes/oep-0001.html#provisional) and the [docs for status changes](https://open-edx-proposals.readthedocs.io/en/latest/processes/oep-0001.html#status-changes), I'd recommend:
1. Updating the status to `Under Review (=> Provisional)` to clarify the intended status upon merge, assuming that status is agreed upon. And,
2. Adding a `Follow-up Work` page to this header as notes in the docs above. This link might also provide some thoughts on how this work will come to be.

2. Additionally, this isn't a blocker to this OEP, but is related to "Follow-up Work". I'm wondering what thought has been given to how companies will need to handle private repos and translations? Maybe a separate Discuss thread could be started for this, but has thought already been given to this? Is the idea that others could easily replicate this pattern and re-use tooling against a private repo of translations? How much work will that involve? Etc.

Author: Thank you for your comments @robrap.

1. If I understand correctly, you want to clarify that the next status after "Under Review" should be "Provisional" and not "Accepted", because it hasn't been vetted and adopted in the platform. This makes sense to me, but I don't think we need to specify this in the document.

2. I will add a link named “Follow-up Work” to the References section of the OEP header.

3. I think this is a good first question for the Follow-up Work link. The openedx-atlas CLI tool can pull translation files from any repository that works with git. And the GitHub Action that generates translations and moves them to openedx-translations can also be reappropriated for private repos. The work involved will be: copying the github action that generates the translations to another repository, changing the organization and names of the repos, and attaching it to a separate Transifex project via the app.

Reviewer 10: @Carlos-Muniz on point 1, @robrap is suggesting that you change the status from:
* `Under Review`

to one of these:

1. `Under Review (=> Provisional)`
2. `Under Review (=> Accepted)`

in order to clarify whether you want this to merge in as a Provisional or Accepted, which is information the reviewers should know. I agree with him.

Reviewer 10: Aside, @robrap : I think this is evidence that the OEP-1 status guidance that we collaborated on could be improved :) OEP-1 presents `Under Review` as a valid status even though we'd really prefer that folks always include the `(=> ...)` part.

Author: Thank you for the clarification. I've changed it.

include being easier to maintain and solving a lot of the pain points detailed below. As
part of this proposal, we suggest moving translations into their own repository, to make
using the `Transifex GitHub App`_ more streamlined and straightforward, and in order to
make organizing and using the up to date translations simpler.

Current State
*************

* edx-transifex-bot is a potential security issue: The edx-transifex-bot requires admin
  rights on Transifex in order to function. Admin rights give access to private/sensitive
  information as well as the ability to permanently delete translation and configuration
  files. At some point, the login to the edx-transifex-bot user was lost, and without
  access to infrastructure that the bot uses to function, this edx-transifex-bot is a
  security issue most of the Open edX community cannot control or debug.
* edx-transifex-bot is a black box for most of the community: the code for the
  edx-transifex-bot is in the `ecommerce-scripts`_ repository but it is impossible for
  most of the community to observe the work it is doing, or whether it is doing it

>   correctly. In addition, there is no documentation for these important scripts.

Reviewer 3: Where are the errors of the Transifex GitHub App shown? I understand that many of the errors will be caught on extraction by the GitHub workflow (and thus will be visible via PR), but I wonder if there are errors that will only be caught at the Transifex GitHub App level.

Assuming such errors exist (they may not):

1. What happens if two projects happen to use the same translation key for a string? (Of course, we have a naming convention to prevent that, but say a copy and paste error?)
2. And will an error caused by one project block translations by all other projects?

Author: I think this is a valid concern. Transifex has a very robust notification system, and errors occurring there will notify the Transifex Admins.

For your thought experiments, Transifex projects are treated exclusively. What happens in one, does not affect the others, similar to how GitHub treats repositories. So an error in any Transifex project other than openedx-translations will not affect the translations in Transifex project. 

As a side note, translation files were kept in their original file structures to combat the non-unique file name errors we have experienced before, as the entire path is used as the unique translation slug. So as long as files are kept in separate directory structures within the openedx-translations repository, this will not be a problem.

Reviewer 3: Ok, great, thanks for the clarification! My sense is that we should be good as long as there is a way for maintainers to know that changes they made caused errors at the Transifex level. It sounds like the Translation WG or Transifex Admins are taking up this role and will be responsible for notifying the relevant maintainer?

I should note that I was being a little confusing in my parent comment because I used "project" to refer to an Open edX Project (a repository) instead of a Transifex Project. I was just trying to make sure an error caused by one repository will not block translations for all other repositories. (I suspect that may have been one of the reasons the original Jenkins jobs that pushed/pulled translations were done on a per-respository basis.)

  there are few people with Admin rights to Transifex and knowledge of the Transifex API;
  this could become a recurring problem with each Open edX release.

.. _ecommerce-scripts: https://github.com/openedx/ecommerce-scripts/tree/master/transifex
.. _this pull request: https://github.com/openedx/edx-platform/pull/30567

Rationale for migrating to the `Transifex GitHub App`_
******************************************************

* This is an upgrade of a system we use regularly, but do not want to have to maintain
  regularly.
* Upgrading from a bot (machine user) to an app/workflow is recommended by GitHub and
  makes the translation process more open source. 
* The `Transifex GitHub App`_ is developed and supported by Transifex
* The `Transifex GitHub App`_ is very simple to configure and has many options. We can
  set Transifex Projects to automatically upload/download translation files from a
  repository once the translations are reviewed and accepted.
* By using an app that is maintained by Transifex the organization, we reduce the
  maintenance burden and are more future proof of changes they might make since they
  maintain both the API and the `Transifex GitHub App`_.

> Rationale for consolidating translations files centrally

Reviewer 2: I'm curious:

(1) What are you asking for?
(2) What "powers" do you want and why?
(3) What "powers" do you currently exercise?

> Rationale for consolidating translations files centrally

Author: The Translation WG is in charge of translating and reviewing translations, and to my knowledge these translations are just automatically merged in, e.g. https://github.com/openedx/paragon/pull/1744. With the changes proposed in OEP-58, Translations WG would still be in charge of translating and reviewing translations. Approval limits (once it hits X% translated it is automatically committed to openedx-translations) can be set per translation file. I can see a situation where maintainers request the Translation WG to set the approval limit below the standard set by the Translations WG. That being said, our Translators are top knotch and keep translations at 90% or higher.

I think answers to @sarina's questions can help shed light to what type of input maintainers are looking for.

> Rationale for consolidating translations files centrally

Reviewer 3: > What are you asking for?

I am trying to understand what the interaction should be between Open edX developers and/or maintainers with the Translation WG. For example, I currently do not know where to find information on the roles and responsibilities of maintainers, developers, and the Translation WG. It may be helpful to point to that information in this OEP.

> What "powers" do you want and why? What "powers" do you currently exercise?

> The Translation WG is in charge of translating and reviewing translations

This is new to me! I greatly appreciate the Translation WG's work, don't get me wrong. However, I thought the status quo was that maintainers ultimately owned their repo's translations. For example, I think many maintainers were routinely given Project Maintainer access to Transifex, and would frequently help add/change/remove strings or review translations. If the ownership of translations is changing or has already changed, it may be helpful to point to that change in this OEP.

Also, previously, maintainers were able to modify their translation files. Now, via this OEP, it appears those translation files are being moved to a centralized repository, and I am unsure who will own that repository and if any of the access maintainers previously had to change those files will change. In short, "will there ever be non-bot PRs for the translation files in openedx-translations?" and "who will approve and/or merge those PRs?".

> Rationale for consolidating translations files centrally

Author: @pshiu brought up some important questions.

- Who will maintain the openedx-translations repository?
- How should errors/bugs in translations in the openedx-translations repository be fixed?
- Who is in charge of approving and/or merging PRs in the openedx-translations repository?

It may be unreasonable to expect any one group to maintain openedx-translations as there are lots of stakeholders. Translators are expected to translate, but may not get to see if their translations will cause errors. Developers may see that there are errors, but may not understand the language of the translation. There will have to be cooperation. 

We should set up an issue request form on openedx-translations (much like [the Github Requests in tcril-engineering](https://github.com/openedx/tcril-engineering/issues/new/choose)) that notifies the Translations WG on slack when there is a translation error that needs to be fixed. This way, Translators can be notified of an error in a way they will be able to digest it, and Developers can write the problem and proposed solution in a way that is common to them.

We will have to look further into how the GitHub Transifex app manages changes to translation files outside of Transifex (https://github.com/openedx/openedx-translations/issues/29) to confirm that making PRs is a possible fix that developers can make. In the case that it is allowed, we can treat it as any other repository and nominate Open edX Community members to be Core Contributors to openedx-translations based on their technical experience and participation in the Translations WG.

@brian-smith-tcril is currently looking at this issue and will update with the Transifex App's capability.

> Rationale for consolidating translations files centrally

Reviewer 4: [i commented on the issue on the openedx-translations repo](https://github.com/openedx/openedx-translations/issues/29#issuecomment-1319105569), but i'll reiterate here:

based on transifex docs https://help.transifex.com/en/articles/6265125-github-via-transifex-ui

> For existing resources that have already been synced with GitHub, if translations are found in GitHub, they will be sent to Transifex only if no translations are found for that language in Transifex

knowing this, it seems we should stick to the idea of transifex as a source of truth for all translated strings, and any updates should happen in transifex

********************************************************

* Transifex only allows a one-to-one relationship between repositories and Transifex
  Projects. Organizing all of the translation files into one repository and one Transifex
  Project has a lower labor cost: projects are managed separately so we end up spending
  less time tracking translation progress, and debugging translation issues when all
  translation files are put in the same place. By decreasing the number of projects we
  need to maintain, we can add more content like the MFE translations.
* A repository that only contains text/binary files, and uses branches to separate
  translations related to Open edX releases can make all interactions with translations
  very quick and simple due to the ability to clone and sparse-checkout the branch of a
  specific release and the directory (repository name) with translation files.

Proposed Implementation
***********************

Move Translation Files to a New Repo
====================================

> Translation files (of types .mo and .po) currently exist amongst the code/documentation

Reviewer 4: ```suggestion
Translation files (of types .mo, .po, and .json) currently exist amongst the code/documentation
```

Author: Good catch!

documentation they translate. We will move these translation files from being amongst the
code/documentation to their own repo. For example, a translation file for the openedx
repository `edx-platform`_ located at
``edx-platform/conf/locale/en/LC_MESSAGES/django.po`` would be moved to the new
repository with the name openedx-translations_ located at

> ``openedx-translation/edx-platform/conf/locale/en/LC_MESSAGES/django.po``. For easier

Reviewer 8: Can I suggest to create an intermediate "repos" folder, such that we have `openedx-translations/repos/edx-platform`? Otherwise we will have a lot of repos located at the root, as we already do: https://github.com/openedx/openedx-translations This will quickly become unwieldy. And if we need extra translations-specific folders at the root, they will be mixed with repo names.

Author: This is a good idea that we can implement once the review period is complete.

reintegration, translation files will be kept in the same directory structure as the
code/documentation they translate.

Repositories that generate translation files will have their translation files generated
and committed via a pull request to the openedx-translation repository via a GitHub

> workflow. Once the translation files from edx-platform and other repositories are moved

Reviewer 3: What will this GitHub workflow be called?

Author: https://github.com/openedx/openedx-translations/blob/main/.github/workflows/extract-translation-source-files.yml

Reviewer 3: Thank you! Would it be worth adding that link into the OEP?

Reviewer 10: @pshiu from [OEP-1](https://open-edx-proposals.readthedocs.io/en/latest/processes/oep-0001.html):
> OEPs are not used to dictate small decisions made in every day feature work. See [OEP-19](https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0019-bp-developer-documentation.html?highlight=documentation#agile-documentation-types) for more detail.

I think the workflow name is a small decision. I wouldn't want to lock Carlos et al into that specific workflow name if it's not critical to this OEP's deisgn. Do you see any harm in leaving it out?

Reviewer 3: Not at all! Thought it might be a nice add if it was already decided for reference.

Author: Since OEP-58's next target status is "Provisional", once it is back "Under Review" with the target status of "Accepted", we can update the details of OEP-58 to reflect the specifics of implementation.

Add `Transifex GitHub App`_ to openedx Organization
===================================================

The `Transifex GitHub App`_ will need to be added to the openedx GitHub organization in
order to grant the app permissions to push/pull the translation files. Currently, we
manage the push/pull permissions for the edx-transifex-bot through a number of GitHub
user groups. The `Transifex GitHub App`_ once installed in an organization, is granted
permissions to push/pull on a repository basis, and by moving all the translation files
to a single repository we eliminate separate translations user groups.

Connect the New Translation Repository to Transifex
===================================================

The Transifex web-app accepts a `Transifex GitHub Integration configuration file`_ for
each Transifex project. By connecting the single repository containing all translation
files, we only need to make a single `Transifex GitHub Integration configuration file`_
that allows the `Transifex GitHub App`_ to manage the translation files. Based on the
Translation Working Group's instruction on acceptable translation/review percentages, we

> can set parameters that automatically push and pull translation files.

Author: Your understanding is correct.

> can set parameters that automatically push and pull translation files.

Reviewer 8: Do you expect any synchronization conflict to occur? If yes, how do you think we will resolve them?

> can set parameters that automatically push and pull translation files.

Reviewer 4: > Do you expect any synchronization conflict to occur? If yes, how do you think we will resolve them?

You raise a very good point! I don't expect sync conflicts to be a common issue, but they're definitely something worth testing for. I made [an issue](https://github.com/openedx/openedx-translations/issues/29) on https://github.com/openedx/openedx-translations for testing/verifying this behavior

> can set parameters that automatically push and pull translation files.

Reviewer 2: > people will be able to make changes to Transifex strings by merging pull requests on the openedx-translations repo. 

I disagree with this as a good idea. The translations teams are largely non-technical and have a process for translating, reviewing, & approving translations. I don't think the maintainers/core contributors to repos have the knowledge or understanding to approve a PR to change translations - and it's not right for someone to overwrite a, say, Spanish translation without going through the es team's review process. Also maintainers don't have write access to Transifex to resolve conflicts. I think the translation process should remain as-is (only translate in Transifex) unless the whole of the Translations Working Group agrees with this change (and understands the ramifications).

> can set parameters that automatically push and pull translation files.

Reviewer 4: >  I think the translation process should remain as-is (only translate in Transifex)

this would definitely simplify things!

> can set parameters that automatically push and pull translation files.

Reviewer 6: I agree @sarina  on the idea that the translations team have the expertise to handle the translations, and that Transifex should be the source of truth for translation strings.

Where things get (technically) difficult is with translations that contain format variables -- Transifex doesn't always flag syntax errors with these, and so they're not noticed until they get pulled and compiled. While I think we can improve our documentation to assist translators with these weirdly formatted strings, we'll still need a process for technical people to fix formatting errors in Transifex which are breaking the translations build.

> can set parameters that automatically push and pull translation files.

Reviewer 3: As an Open edX developer, I've occasionally have had valid reason to make edits to translation files. Here are a couple example `git log` commands that will show commits by non-bot users to translation files:

```
Developer % cd edx-platform
edx-platform % git log --stat -P --author '^((?!sre\+jenkins|edx-semantic-release|edx-transifex-bot|learner-engineering).)*$' -- conf/locale/**/*.po
edx-platform % cd ../frontend-app-learning
frontend-app-learning % git log --stat -P --author '^((?!sre\+jenkins|edx-semantic-release|learner-engineering).)*$' -- src/i18n/messages
frontend-app-learning % cd ../frontend-app-profile
frontend-app-profile % git log --stat -P --author '^((?!sre\+jenkins|edx-semantic-release|edx-transifex-bot|learner-engineering).)*$' -- src/i18n/messages
```

> can set parameters that automatically push and pull translation files.

Reviewer 2: The correct way for these fixes - and the correct way as requested by the translators WG - is to ask them to fix the problem. They can be found at https://openedx.slack.com/archives/C037XDB9KN1

I think in your case the fix should be the quick fix on the translation file AND an immediate ping to the translation teams to fix it canonically in Transifex.

> can set parameters that automatically push and pull translation files.

Reviewer 2: Pushing changes back to transifex is just a bad idea. What if you make a change to a translation and in the meantime the translators have decided to completely change how the string is translated? Then you'll overwrite their changes and they won't know.

> can set parameters that automatically push and pull translation files.

Author: Without access to the jenkins server running the edx-transifex-bot, I'm not sure if there is a way for the majority of the community to know what the percentage thresholds are or if there is another heuristic to when translation files are downloaded from Transifex back to GitHub.

> can set parameters that automatically push and pull translation files.

Reviewer 3: It looks like 3%! So I am no longer concerned. Thank you. Here's the command from i18n-tools [i18n/transifex.py](https://github.com/openedx/i18n-tools/blob/2b229f5ceb42e9dc52a84c37b07dea32f4a8be86/i18n/transifex.py#L57):
```
 cmd = f'tx pull -t -f --mode=reviewed --minimum-perc=3 -l {lang}'
```
Here's the trace to i18n-tools:
* ecommerce-scripts [transifex/utils.py](https://github.com/openedx/ecommerce-scripts/blob/563b4d6d9185961412d1b20ee1d5c5348e4ba361/transifex/utils.py#L192)
* edx-platform [Makefile](https://github.com/openedx/edx-platform/blob/9f5596c0af9386a2dbe8ca7019f6b4a8fd458583/Makefile#L48)

> can set parameters that automatically push and pull translation files.

Author: https://github.com/openedx/open-edx-proposals/pull/367/files#r1024179984

translators and reviewers across Transifex projects into the new project associated with
the new repository. In addition, we can save all the progress the Open edX translators
have accomplished by copying the Transifex's Translation Memory, the auto-translation
feature that allows for Projects with similar strings to be automatically translated,
from the old projects to this new one. Once older projects are made redundant by the new
project, they will be deprecated. By moving all the translation files to the same
repository we can increase the reach of the Transifex's Translation Memory feature to
help translate similar strings across the entire code/documentation base.

Get Translations Back for Deployment/Development
================================================

A new python library, called openedx-atlas, will be created. This will enable the
placement of the translation files kept in openedx-translations into locally cloned
repositories for development and containers containing the code translation files are
formed from. This tool will manage the placement of translation files through an editable

> atlas configuration file (atlas.yml) kept in the repositories that have

Author: openedx-atlas already provides the ability to override default parameter values and an atlas.yml configuration file by using the CLI's flags. 

`atlas pull -r "openedx/openedx-translations" -b "main" -d "credentials"`

This would pull translation files from the credentials directory in the main branch 
 from the repository openedx/openedx-translations.

There are many other features that could be beneficial, such as the ones you described in your comment. We can hash out these implementation details further in [Follow-up Work](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3578494977/OEP-58+Translations+Management).

> atlas configuration file (atlas.yml) kept in the repositories that have

Reviewer 10: semi-related thread: https://github.com/openedx/open-edx-proposals/pull/367#discussion_r1023212567

translation files kept in openedx-translations. The atlas.yml file will support
options that allow for the concatenation, reorganization, and reformatting of translation
files as they are copied to their locations amongst the code. The atlas.yml file
will also support selecting which languages to be included in an Open edX deployment. The
tool will have to be used/ran as part of the setup of a repository, whether for

> development or deployment.

Author: Since openedx-atlas works as a configurable wrapper to git's sparse-checkout, openedx-atlas can be used to pull translations from openedx-translations right after git clone is used to pull the repository from GitHub. This should cover most supported development & deployment environments, though I'm curious if there are any others that may not work like this.

> development or deployment.

Reviewer 4: > This should cover most supported development & deployment environments, though I'm curious if there are any others that may not work like this.

Having atlas built in python adds complexity to the deployment of MFEs. It seems there are a few paths that we could go down

The first option would be to keep atlas in python, and to use the python implementation everywhere
```
Pros:
* Flexibile/Extensibile
* Easy to maintain (only need to maintain one implementation)
Cons:
* Requires python in build/deployment envs
```

Another option would be to write atlas in bash, that way python would not be required in build/deployment environments for MFEs.

```
Pros:
* Runs (pretty much) everywhere
Cons:
* Bash can get messy when projects get complex
```

A third option would be to *also* write atlas functionality in js and package for NPM so MFEs (or frontend-build) could import it.

```
Pros:
* Easy integration for MFEs
* Flexible/extensible
Cons:
* Requires maintaining duplicated functionality (need to implement everything in both js and python)
```

If the scope of atlas stays limited, i'd lean towards the bash option. If there's an expectation of increased complexity within atlas, however, there are definite benefits to using full fledged programming languages.

> development or deployment.

Reviewer 10: @bmedx mentioned a fourth option: get rid of openedx-atlas, and publish releases of .tar.gz/.zip files from the openedx-translations repository
```
Pros:
* No sparse checkout logic needed, no separate CLI tool needed.
* All the complexity would be in the openedx-translations repo; other repos would just download & unpack the .tar.gzs.
Cons:
* Would need to add automation to the openedx-translations repo to regularly publish these
* The .tar.gzs would contain every language (or every supported language?)
  * It's unclear whether the .tar.gzs would be too big or not.
     edx-platform's locale directory is currently 60 Mb, gzips down to 14 Mb.
* Is there other stuff we think openedx-atlas would do that we'd be leaving out with this option?
```

> development or deployment.

Reviewer 11: Solely from the MFE repo perspective, I'm for gzips.  Github workflows for publishing them on every commit would be pretty simple.  And I very much doubt download sizes will be problematic.  For all I know `tx pull` could be even less efficient.  🤷🏼 

What would stop us from storing translations per-repo in opendx-translations, though?  That would make each individual pull pretty tiny.

If that doesn't work, for whatever reason, I'd vote for bash.

> development or deployment.

Reviewer 10: @arbrandes 

> What would stop us from storing translations per-repo in opendx-translations, though? That would make each individual pull pretty tiny.

That is indeed the suggestion on the table -- one gzip published per repo 🙂 It's just that each gzip would have every language *unless* we wanted to publish R*L gzips, where R is the number of repos and L is the number of languages.

Impacts
*******

Impact on Translators
=====================

As we approach the end of the translation upgrade process, we will need to tactically
move from multiple Transifex projects to a single project. This will require coordination
with our translators to ensure that moving forward they are providing translations in the
right place.

Impact on Site Operators
========================

Currently the translation files for any given service or library is stored at the same
place as the code, which has generally simplified the deployment story in the past. With
this change, the translations files will move to their own repository. As we deprecate
the old translations files, the relevant deployment tooling will need to be updated to
pull down the translations from the new repository as a part of the deployment process.
This will impact both the old Ansible based tooling as well as any new Docker based
tooling.

Impact on Developers
====================

While it won’t directly impact the day-to-day workflow of developers (unless you are

> developing or testing with translation files), due to the same reasons that we impact

Author: Learning how and when to use openedx-atlas may be a non-negligible learning task, but then again so is learning any new package. In the future, it can be added to Makefiles to minimize the learning load just as i18n-tools commands were.

> developing or testing with translation files), due to the same reasons that we impact

Reviewer 4: I get the impression that for most dev workflows the atlas usage will be pretty straightforward. Any friction can also be mitigated by ensuring repos include a section in the readme describing any simple atlas commands a dev may need to run. Furthermore, atlas use can be integrated into any scripts that currently depend on translation files existing within the repo.

> developing or testing with translation files), due to the same reasons that we impact

Reviewer 3: @brian-smith-tcril, I really like the README addition idea and integrating openedx-atlas into any scripts we currently use within our repos.

I definitely agree the impact will be minimal; it's clear openedx-atlas was designed with that in mind. But minimal impact does exist, and I think developer education is a great way to address that impact.

project already exist in a repository with the name of openedx/openedx-i18n. A new
repository named openedx/openedx-translations will be similarly structured, but it will
contain the translation files for all repositories within openedx. The
`Transifex GitHub App`_ will be installed in the openedx organization. Similar to how the
Build-Test-Release Working Group creates a new branch for each new named release of
edx-platform, translation releases will also be kept in branches corresponding to
edx-platform releases.

Rejected Alternatives
*********************

Rewriting the Current Tooling for the New API
=============================================

The source code for the edx-transifex-bot can be found in `ecommerce-scripts`_. We could
rewrite the current tooling to try to solve the problems encountered in the last two Open
edX releases and upgrade to the new API, but this approach is a patch-up job that will
not address several other issues mentioned and would have to be undertaken by the
community member with exclusive access to the legacy infrastructure currently running the
edx-transifex-bot.

> Making a Transifex Project for Each Repository

Author: Possibly, but then we have to consider that a mostly non-technical Translation WG has to learn how to use the Transifex API. And then Translators may have to be part of 100+ teams to translate files for each repository, and learn to navigate the hundreds of projects to see their progress.

> Making a Transifex Project for Each Repository

Reviewer 2: @pshiu for this and some other comments, it would be useful to spell out your use cases (both ones you currently follow and new ones you're proposing - and for the latter, justification to the need)

> Making a Transifex Project for Each Repository

Reviewer 3: @sarina Sorry for the confusion! My comment was meant to address the rejected alternative of making a Transifex project for each repository. The suggestion was that perhaps tooling could be created to remedy the concerns that caused this alternative be rejected. (For example, perhaps a bot that auto-adds and removes translators from all relevant projects.)

@Carlos-Muniz I thought that even in a consolidated repository scenario, translators still have to navigate one [Transifex Resource](https://help.transifex.com/en/articles/6656105-introduction-to-projects-and-resources) per original source repository (resulting in the same problem)?

> Making a Transifex Project for Each Repository

Author: @pshiu Translators are capable of navigating between Transifex Resources (Translation files) within one Transifex Project, as they do that already with the current Translation Project "edx-platform". File structures are flattened within a transifex project, so they will not have to search through directories to find the resources they are to translate. But having to look between 100+ Transifex projects for 1-2 resources each project, and then expecting Transifex Admin to be able to track the progress of each of the languages for each of the projects becomes a far more labor intensive alternative. 

Tooling + a dashboard to track progress would have to be made at the very least. And with all those extra moving parts, we may find ourselves with a pile of spaghetti that would be very difficult to untangle.

> Making a Transifex Project for Each Repository

Reviewer 3: @Carlos-Muniz After diving into Transifex's UI, here's what I now understand:
* Transifex's string search [[example search](https://www.transifex.com/open-edx/search/?q=translation_text%3Abuscar+target_language%3Aes_419), [docs](https://help.transifex.com/en/articles/6240799-search-strings)] allows for cross-project string searches.
* Dashboards that document language progress across projects [[all projects dashboard](https://www.transifex.com/open-edx/), [all projects language dashboard](https://www.transifex.com/open-edx/dashboard/all_projects/es_419/)] seem to be available.
* The Transifex editor supports an "All Resources" option, but not an "All Projects" option.

That last bullet means (like you said) that to translate or review a string across projects, a translator will either have to:

1. Click on the editor links in the string search, which loads an editor in a new tab, or
2. Know which project a string belongs to and switch to that project in the editor's project dropdown menu.

Both are tedious for the translator and a bad user experience. The fact that the Transifex editor does not support editing strings across multiple projects is (at least for me) the winning argument justifying the consolidation of all translation strings into a single Transifex project.

I didn't understand this when I first read the OEP, so I really appreciate your time in enlightening me. Thanks for your help.

OEP-58: Translations Management
###############################

.. list-table::
   :widths: 25 75

   * - OEP
     - :doc:`OEP-58 <oep-0058-arch-translations-management>`
   * - Title
     - Translations Management
   * - Last Modified
     - 2022-11-07
   * - Authors
     - 
       * Carlos Muniz <cmuniz@tcril.org>
       * Feanil Patel <feanil@tcril.org>
       * Sarina Canelake <sarina@tcril.org>
   * - Arbiter
     - Ned Batchelder <ned@edx.org>
   * - Status
     - Under Review
   * - Type
     - Architecture Decision
   * - Created
     - 2022-08-08
   * - Review Period
     - 2022-11-07 - 2022-11-21
   * - References
     - `Follow-up Work <https://openedx.atlassian.net/wiki/spaces/COMM/pages/3578494977/Translations+Management+Design+Implementation>`_
..    * - Resolution
..      - 

.. contents::
  :local:
  :depth: 1

> Context

Author: Thank you for this suggestion. I will reorganize the OEP this way at the end of the review period so comments on specific sections are referenced correctly by GitHub.

development tools as well. In addion, we will create new instructions for developers on
how to enable translations for a new service/repo when it comes online.

Final State
***********

Repositories that generate translation files will have their translation files generated

> and pushed to the openedx-translation repository via a github workflow. Once the

Reviewer 12: "pushed" here means committed via a pull request?

Author: Yes. I'll change this so its clearer.

translation files from edx-platform and other repositories are moved to the
openedx-translations repository, the Transifex GitHub App will link a Transifex project
of a name such as "Open edX Translations" to the openedx-translations repository. A

> configuration file naming the files that are to be translated and the trigger that pulls

Reviewer 12: This configuration file is used by the Transifex GitHub App?  Do you have a link for the app, and the configuration it allows?

Author: Here is the [Transifex GitHub App](https://github.com/apps/transifex-integration), and the [configuration](https://docs.transifex.com/transifex-github-integrations/github-tx-ui#linking-a-specific-project-with-a-github-repository) it allows. I'll add them to the OEP too.

translation files back into will be created in the openedx/translations repository. This
link will allow for the Transifex GitHub App to automatically manage the push/pull of the
translation files without the need for human intervention.

When it comes time to cut an Open Release, a new branch will be formed in the
openedx-translations repository for this release. This new branch will correspond with
other release branches such as those found in edx-platform. It is my recommendation that
a tool in the form of a python library is written to enable the placement of the
translation files kept in openedx-translations into the repositories the translation

> files are formed from. This tool will manage the placement of translation files through

Reviewer 12: It sounds like this tool would be used at release time? But doesn't it need to do the placement and concatenation o of translation files for every deploy?

Reviewer 12: Oh, you said that below.
